projects/core/src/lib/mask.ts

```diff
@@ -50,34 +53,72 @@ export class Mask {
             return;
         }
 
-        const currentValue = this.elementRef.value;
-        const caretPosition = this.elementRef.selectionStart || 0;
-        const newPossibleValue =
-            currentValue.slice(0, caretPosition) +
-            event.key +
-            currentValue.slice(caretPosition);
+        const {value, selectionStart, selectionEnd} = this.elementRef;
+        const maskModel = new MaskModel(value, selectionStart || 0, this.maskExpression);
 
-        if (!validateValueWithMask(newPossibleValue, this.maskExpression)) {
+        try {
+            maskModel.addCharacters([selectionStart || 0, selectionEnd || 0], pressedKey);
+        } catch {
             event.preventDefault();
         }
     }
 
+    private fillWithFixedValues(): void {
+        const {value: initialValue, selectionStart} = this.elementRef;
+        const maskModel = new MaskModel(
+            initialValue,
+            selectionStart || 0,
+            this.maskExpression,
+        );
+        const {value, caretIndex} = maskModel;
+
+        this.updateValue(value);
+        this.updateCaretIndex(caretIndex);
+    }
+
+    private handleBackspace(event: KeyboardEvent): void {
+        const {maskExpression} = this;
+
+        if (!Array.isArray(maskExpression)) {
+            return;
+        }
+
+        event.preventDefault();
+
+        const {selectionStart, selectionEnd, value: initialValue} = this.elementRef;
+        const maskModel = new MaskModel(
+            initialValue,
+            selectionStart ?? 0,
+            maskExpression,
+        );
+
+        maskModel.removeCharacters([selectionStart ?? 0, selectionEnd ?? 0]);
+
+        const {value, caretIndex} = maskModel;
+
+        this.updateValue(value);
+        this.updateCaretIndex(caretIndex);
+    }
+
     private handlePaste(event: ClipboardEvent): void {
         // TODO: finish later
         // eslint-disable-next-line no-console
         console.log('===[handlePaste]===', event.cancelable, getClipboardDataText(event));
     }
 
-    private fillInputWithFixedValues(isBackspace: boolean = false): void {
-        const {value} = this.elementRef;
-        const modifiedValue = adjustFixedMaskCharacters(
-            value,
-            this.maskExpression,
-            isBackspace,
-        );
+    private updateValue(newValue: string): void {
+        const {elementRef} = this;
+
+        if (elementRef.value !== newValue) {
+            elementRef.value = newValue;
+        }
+    }
+
+    private updateCaretIndex(newCaretIndex: number): void {
+        const {elementRef} = this;
```

Comment:
And here :) I have a feeling that less lines would be easier to scan and follow.